integration/e2e_test.go

```diff
@@ -171,120 +175,63 @@ func deployCluster(t testing.TB, cluster_manifest_file string) {
 	t.Logf("Cluster name is %s", cluster)
 
-	attempt := 0
-	maxAttempts := 60
 	command = []string{"clusterctl", "get", "kubeconfig", cluster}
-	for {
-		cmd = exec.Command(command[0], command[1:]...)
-		outputBytes, err = cmd.Output()
+	testFn := func(output string) bool {
+		cfg := strings.Trim(output, "\n")
+		err = os.WriteFile(KUBECONFIG, []byte(cfg), 0644)
 		if err != nil {
-			if attempt >= maxAttempts {
-				t.Fatal(err)
-			} else {
-				attempt++
-				t.Log("Failed to get the target's kubeconfig, retrying.")
-				time.Sleep(20 * time.Second)
-			}
-		} else {
-			cfg := strings.Trim(string(outputBytes), "\n")
-			err = os.WriteFile(KUBECONFIG, []byte(cfg), 0644)
-			if err != nil {
-				t.Fatalf("Could not persist the targets kubeconfig file. %s", err)
-			}
-			t.Logf("Target's kubeconfig file is at %s", KUBECONFIG)
-			t.Log(cfg)
-			break
+			t.Fatalf("Could not persist the targets kubeconfig file. %s", err)
 		}
+		t.Logf("Target's kubeconfig file is at %s", KUBECONFIG)
+		t.Log(cfg)
+		return true
+	}
+	if err = retryableCommandWithConditionFunction(t, command, testFn); err != nil {
+		t.Fatal(err)
 	}
 
 	// Wait until the cluster is provisioned
-	attempt = 0
-	maxAttempts = 60
 	command = []string{"kubectl", "get", "cluster", cluster}
-	for {
-		cmd = exec.Command(command[0], command[1:]...)
-		outputBytes, err = cmd.CombinedOutput()
-		if err != nil {
-			t.Log(string(outputBytes))
-			if attempt >= maxAttempts {
-				t.Fatal(err)
-			} else {
-				attempt++
-				t.Log("Retrying")
-				time.Sleep(10 * time.Second)
-			}
-		} else {
-			if strings.Contains(string(outputBytes), "Provisioned") {
-				break
-			} else {
-				attempt++
-				time.Sleep(20 * time.Second)
-				t.Log("Waiting for the cluster to be provisioned")
-			}
-		}
+	testFn = func(output string) bool {
+		return strings.Contains(output, "Provisioned")
+	}
+	if err = retryableCommandWithConditionFunction(t, command, testFn); err != nil {
+		t.Fatal(err)
 	}
+}
 
+// verifyCluster check if cluster is functional
+func verifyCluster(t testing.TB) {
 	// Wait until all machines are running
-	attempt = 0
-	maxAttempts = 60
+
 	machines := 0
-	command = []string{"kubectl", "get", "machine", "--no-headers"}
-	for {
-		cmd = exec.Command(command[0], command[1:]...)
-		outputBytes, err = cmd.CombinedOutput()
-		output := string(outputBytes)
-		if err != nil {
-			t.Log(output)
-			if attempt >= maxAttempts {
-				t.Fatal(err)
-			} else {
-				attempt++
-				t.Log("Retrying")
-				time.Sleep(10 * time.Second)
-			}
-		} else {
-			machines = strings.Count(output, "\n")
-			running := strings.Count(output, "Running")
-			t.Logf("Machines %d out of which %d are Running", machines, running)
-			if machines == running {
-				break
-			} else {
-				attempt++
-				time.Sleep(10 * time.Second)
-				t.Log("Waiting for machines to start running")
-			}
+	command := []string{"kubectl", "get", "machine", "--no-headers"}
+	testFn := func(output string) bool {
```

Comment:
@bschimke95 is that refactor method ok?

Reply:
the retry should not depend on the context (e.g. executing a command in this case).
Implement something like this:
https://github.com/canonical/k8s-snap/blob/1b51ba74e57089eb8362d4f99452103381849028/src/k8s/pkg/utils/control/retry.go
or this if you want to wait for a condition:
https://github.com/canonical/k8s-snap/blob/1b51ba74e57089eb8362d4f99452103381849028/src/k8s/pkg/utils/control/wait.go